dependencies/src/main/java/com/google/cloud/tools/opensource/classpath/LinkageCheckerMain.java

```diff
@@ -145,4 +77,102 @@ public static void main(String[] arguments)
+      linkageProblems =
+          linkageProblems.stream()
+              .filter(
+                  (LinkageProblem problem) ->
+                      graph.isReachable(problem.getSourceClass().getBinaryName()))
+              .collect(toImmutableSet());
+    }
+
+    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
+    if (writeAsExclusionFile != null) {
+      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
+      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
+      return ImmutableSet.of();
+    }
+
+    if (!linkageProblems.isEmpty()) {
+      System.out.println(LinkageProblem.formatLinkageProblems(linkageProblems));
+    }
+
+    return linkageProblems;
+  }
+
+  private static ImmutableSet<LinkageProblem> runWithArtifacts(
+      LinkageCheckerArguments linkageCheckerArguments)
+      throws IOException, RepositoryException, TransformerException, XMLStreamException {
+    ImmutableList<Artifact> artifactsInArguments = linkageCheckerArguments.getArtifacts();
+
+    // When a BOM or Maven artifacts are passed as arguments, resolve the dependencies.
+    DependencyGraphBuilder dependencyGraphBuilder =
+        new DependencyGraphBuilder(linkageCheckerArguments.getMavenRepositoryUrls());
+    ClassPathBuilder classPathBuilder = new ClassPathBuilder(dependencyGraphBuilder);
+    ClassPathResult classPathResult = classPathBuilder.resolve(artifactsInArguments, false);
+    ImmutableList<ClassPathEntry> inputClassPath = classPathResult.getClassPath();
+    ImmutableList<ArtifactProblem> artifactProblems =
+        ImmutableList.copyOf(classPathResult.getArtifactProblems());
+    ImmutableSet<ClassPathEntry> entryPoints =
+        ImmutableSet.copyOf(inputClassPath.subList(0, artifactsInArguments.size()));
+
+    LinkageChecker linkageChecker =
+        LinkageChecker.create(
+            inputClassPath, entryPoints, linkageCheckerArguments.getInputExclusionFile());
+    ImmutableSet<LinkageProblem> linkageProblems = linkageChecker.findLinkageProblems();
+
+    if (linkageCheckerArguments.getReportOnlyReachable()) {
+      ClassReferenceGraph graph = linkageChecker.getClassReferenceGraph();
+      linkageProblems =
+          linkageProblems.stream()
+              .filter(
+                  (LinkageProblem problem) ->
+                      graph.isReachable(problem.getSourceClass().getBinaryName()))
+              .collect(toImmutableSet());
+    }
+
+    LinkageProblemCauseAnnotator.annotate(classPathBuilder, classPathResult, linkageProblems);
+
+    Path writeAsExclusionFile = linkageCheckerArguments.getOutputExclusionFile();
+    if (writeAsExclusionFile != null) {
+      ExclusionFiles.write(writeAsExclusionFile, linkageProblems);
+      System.out.println("Wrote the linkage errors as exclusion file: " + writeAsExclusionFile);
+      return ImmutableSet.of();
+    }
+
+    if (!linkageProblems.isEmpty()) {
```

Comment:
tempted to move this if block to a printLinkageProblems method too

Reply:
I searched the usage of `formatLinkageProblems`.  The enforcer rule uses logger. That `printLinkageProblems` method would be only used by LinkageCheckerMain. So it's better to have if statement here.